ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala

```diff
@@ -527,43 +618,50 @@ private object OracleQueries extends Queries {
 
   protected[this] type DBContractKey = JsValue
 
-  protected[this] override def toDBContractKey[CK: JsonWriter](x: CK) =
+  protected[this] override def toDBContractKey[CK: JsonWriter](x: CK): DBContractKey =
     JsObject(Map("key" -> x.toJson))
 
   protected[this] override def primInsertContracts[F[_]: cats.Foldable: Functor](
       dbcs: F[DBContract[SurrogateTpId, DBContractKey, JsValue, Array[String]]]
   )(implicit log: LogHandler, pas: Put[Array[String]]): ConnectionIO[Int] = {
     println("insert contracts")
     println(dbcs)
-    val r = Update[(String, SurrogateTpId, JsValue, JsValue, String)](
+    val r = Update[(String, String, SurrogateTpId, JsValue, JsValue, String)](
       """
-        INSERT INTO contract(contract_id, tpid, key, payload, agreement_text)
+        MERGE INTO contract USING (SELECT 1 FROM DUAL) ON (contract_id = ?)
```

Comment:
Assuming the `MERGE INTO` stuff is to handle `on conflict do nothing`, I noticed that you can do that a bit easier https://github.com/digital-asset/daml/commit/2c14ff1e80538883156aacbb04ee2798ebfa8092#

Reply:
Nice. I'll incorporate that into the next PR, though.

Reply:
e6a53be22a7b34c233110d63d1e5cd20cd87d50b